### PR TITLE
Fix storage merge on view cannot use primary index

### LINKED ISSUE #40014: Merge Table Engine does not use the indexes of tables.
We use Merge Table Engine to merge a ReplicatedMergeTree Engine table and a normal view without Distributed Table Engine based on another ReplicatedMergeTree Engine table.

And we found that the Merge Table Engine does not use the indexes of view.

But if we use Distributed Table Engine to make these tables and views distributed, the Merge Table Engine will use their indexes.

So what is the difference between these two modes?

ClickHouse version: 21.9.2.17

### PR DESCRIPTION
**Pull request by canhld94**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Fix storage merge on view cannot use index

Closes  #40014 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by den-crane** [comment]
@canhld94 you should use `currentDatabase()`
 and you can use `force_primary_key=1` to check is it use Primary index or not

CREATE TABLE m2 (id UInt64) ENGINE=Merge(currentDatabase(),'m0|m1v');
SELECT * FROM m2 WHERE id > 1 AND id < 5 settings force_primary_key=1, max_bytes_to_read = 64;


```sql
20.3.21.2

CREATE TABLE m0 (id UInt64) ENGINE=MergeTree ORDER BY id SETTINGS index_granularity = 1;
INSERT INTO m0 SELECT number FROM numbers(10);
CREATE TABLE m1 (id UInt64, s String) ENGINE=MergeTree ORDER BY id SETTINGS index_granularity = 1;
INSERT INTO m1 SELECT number, 'boo' FROM numbers(10);
CREATE VIEW m1v AS SELECT id FROM m1;

CREATE TABLE m2 (id UInt64) ENGINE=Merge(currentDatabase(),'m0|m1v');

SELECT * FROM m2 WHERE id > 1 AND id < 5 settings force_primary_key=1, max_bytes_to_read = 64;
┌─id─┐
│  2 │
│  3 │
│  4 │
└────┘
┌─id─┐
│  2 │
│  3 │
│  4 │
└────┘
```

```sql
22.7.1.2484
SELECT * FROM m2 WHERE id > 1 AND id < 5 settings force_primary_key=1, max_bytes_to_read = 64;

Received exception from server (version 22.7.1):
Code: 277. DB::Exception: Received from localhost:9000. DB::Exception: Primary key (id) is not used and setting 
'force_primary_key' is set. (INDEX_NOT_USED)
```

**Comment by canhld94** [comment]
> @canhld94 you should use `currentDatabase()` -- `Merge(currentDatabase(),'m0|m1v');` (the current database is not `test` your ENGINE=Merge won't work.
> 
>  and you can use `force_primary_key=1` to check is it use Primary index or not
> 
> 
> 
> ```sql
> 
> 20.3.21.2
> 
> 
> 
> CREATE TABLE m0 (id UInt64) ENGINE=MergeTree ORDER BY id SETTINGS index_granularity = 1;
> 
> INSERT INTO m0 SELECT number FROM numbers(10);
> 
> CREATE TABLE m1 (id UInt64, s String) ENGINE=MergeTree ORDER BY id SETTINGS index_granularity = 1;
> 
> INSERT INTO m1 SELECT number, 'boo' FROM numbers(10);
> 
> CREATE VIEW m1v AS SELECT id FROM m1;
> 
> 
> 
> CREATE TABLE m2 (id UInt64) ENGINE=Merge(currentDatabase(),'m0|m1v');
> 
> 
> 
> SELECT * FROM m2 WHERE id > 1 AND id < 5 settings force_primary_key=1, max_bytes_to_read = 64;
> 
> ┌─id─┐
> 
> │  2 │
> 
> │  3 │
> 
> │  4 │
> 
> └────┘
> 
> ┌─id─┐
> 
> │  2 │
> 
> │  3 │
> 
> │  4 │
> 
> └────┘
> 
> ```
> 
> 
> 
> ```sql
> 
> 22.7.1.2484
> 
> SELECT * FROM m2 WHERE id > 1 AND id < 5 settings force_primary_key=1, max_bytes_to_read = 64;
> 
> 
> 
> Received exception from server (version 22.7.1):
> 
> Code: 277. DB::Exception: Received from localhost:9000. DB::Exception: Primary key (id) is not used and setting 
> 
> 'force_primary_key' is set. (INDEX_NOT_USED)
> 
> ```

Updated

**Comment by den-crane** [comment]
@canhld94 it seems your changes broke the test https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/00270_views_query_processing_stage.sql

**Comment by canhld94** [comment]
> @canhld94 it seems your changes broke the test https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/00270_views_query_processing_stage.sql

I located the issue, will fix when I have time tomorrow.

File: tests/queries/0_stateless/02402_merge_engine_with_view.reference [hunk 1]
```diff
@@ -0,0 +1,6 @@
+2
```
Comment (nickitat) [inline tests/queries/0_stateless/02402_merge_engine_with_view.reference:1]: does something guarantee this particular order of numbers in output? if no, pls add `order by` to your query

File: src/Storages/StorageMerge.cpp [hunk 2]
```diff
@@ -528,15 +529,33 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
...
-            UInt32(streams_num));
+        if (StorageView * view = dynamic_cast<StorageView *>(storage.get()))
+        {
+            /// For view storage, we need to rewrite the `modified_query_info.view_query` to optimize read.
+            /// The most intuitive way is to use InterpreterSelectQuery.
+
+            /// Intercept the settings
+            modified_context->setSetting("max_threads", streams_num);
+            modified_context->setSetting("max_streams_to_max_threads_ratio", 1);
+            modified_context->setSetting("max_block_size", max_block_size);
```
Comment (nickitat) [inline src/Storages/StorageMerge.cpp:540]: why do we want to tweak these settings?

File: tests/queries/0_stateless/02402_merge_engine_with_view.reference (hunk 1 above)
Comment (canhld94) [reply tests/queries/0_stateless/02402_merge_engine_with_view.reference:1]: I will add it

File: src/Storages/StorageMerge.cpp (hunk 2 above)
Comment (canhld94) [reply src/Storages/StorageMerge.cpp:540]: Just to respect the logic of old code. In the end we will read from a real storage where these settings applied.

**Comment by canhld94** [comment]
Fail tests look unrelate cc @nickitat

**Review by nickitat — APPROVED** [review]

**Pull request merged**